### PR TITLE
SNOW-893835 Arrow - support for multi-chunk results

### PR DESCRIPTION
**Pull request by sfc-gh-dstempniak**
### Description
Arrow - support for multi-chunk results.

### Checklist
- [x] Code compiles correctly
- [x] Code is formatted according to [Coding Conventions](../CodingConventions.md)
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing (`dotnet test`)
- [x] Extended the README / documentation, if necessary
- [x] Provide JIRA issue id (if possible) or GitHub issue id in PR name

### CODE REVIEW

File: Snowflake.Data/Core/SFBlockingChunkDownloaderV3.cs [hunk 1]
```diff
@@ -64,18 +65,26 @@ class SFBlockingChunkDownloaderV3 : IChunkDownloader
...
-            this.taskQueues = new List<Task<IResultChunk>>();
+            this.taskQueues = new List<Task<BaseResultChunk>>();
             externalCancellationToken = cancellationToken;
 
             for (int i=0; i<prefetchSlot; i++)
             {
-                SFReusableChunk reusableChunk = new SFReusableChunk(colCount);
-                reusableChunk.Reset(chunkInfos[nextChunkToDownloadIndex], nextChunkToDownloadIndex);
-                chunkDatas.Add(reusableChunk);
+                BaseResultChunk resultChunk;
```
Comment (sfc-gh-pfus) [inline Snowflake.Data/Core/SFBlockingChunkDownloaderV3.cs:73]: Can we use ternary operator to init here?

File: Snowflake.Data.Tests/UnitTests/ArrowResultChunkTest.cs [hunk 2]
```diff
@@ -27,38 +29,135 @@ public void SFArrowResultChunkTestDone()
...
+                _chunk.Next();
+                Assert.AreEqual(valueFromRecordBatch, _chunk.ExtractCell(0).SafeToString());
             }
         }
 
         [Test]
         public void TestExtractCellThrowsOutOfRangeException()
         {
-            Assert.Throws<ArgumentOutOfRangeException>(() => _chunk.ExtractCell(RowCount, 0).SafeToString());
+            for (var i = 0; i < RowCountBatchOne; ++i)
```
Comment (sfc-gh-mhofman) [inline Snowflake.Data.Tests/UnitTests/ArrowResultChunkTest.cs:142]: simply: while (_chunk.Next()); ?

File: Snowflake.Data.Tests/UnitTests/ArrowResultChunkTest.cs [hunk 3]
```diff
@@ -27,38 +29,135 @@ public void SFArrowResultChunkTestDone()
...
+        }
+
+        [Test]
+        public void TestRewindIteratesThroughAllRecordsOfOneBatch()
+        {
+            for (var i = 0; i < RowCountBatchOne; ++i)
+            {
+                _chunk.Next();
+            }
+            _chunk.Next();
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply Snowflake.Data.Tests/UnitTests/ArrowResultChunkTest.cs:82]: If I understand correctly this return false. It is needed to rewind to the end of previous chunk?

File: Snowflake.Data.Tests/UnitTests/ArrowResultSetTest.cs [hunk 4]
```diff
@@ -92,25 +92,34 @@ public void TestHasRowsReturnsFalseIfNoRows()
...
         public void TestRewindReturnsTrueForThirdRowAndMovesToFirstRow()
         {
-            _arrowResultSet.Next();
-            _arrowResultSet.Next();
+            _arrowResultSet.Next(); // move to first row
+            _arrowResultSet.Next(); // move to second row
+            _arrowResultSet.Next(); // move to third row
             Assert.IsTrue(_arrowResultSet.Rewind());
             Assert.IsTrue(_arrowResultSet.Rewind());
             Assert.IsFalse(_arrowResultSet.Rewind());
```
Comment (sfc-gh-mhofman) [inline Snowflake.Data.Tests/UnitTests/ArrowResultSetTest.cs:125]: cannot place comment below - should we do sth about ln 152 or can we get rid of the comment?

**Review by sfc-gh-mhofman — CHANGES_REQUESTED** [review]

**Review by sfc-gh-mhofman — APPROVED** [review]

File: Snowflake.Data.Tests/IntegrationTests/SFDbCommandIT.cs [hunk 5]
```diff
@@ -238,13 +238,15 @@ public void TestSimpleLargeResultSet()
...
-                while (reader.Read())
+                cmd.CommandText = "select seq4(), uniform(1, 10, 42), UUID_STRING() as Pierwsz_kolumna, UUID_STRING() as Druga_kolumna from table(generator(rowcount => 5000)) v order by 1";
+                using (IDataReader reader = cmd.ExecuteReader())
                 {
-                    Assert.AreEqual(counter.ToString(), reader.GetString(0));
-                    counter++;
+                    int counter = 0;
+                    while (reader.Read())
+                    {
+                        Assert.AreEqual(counter.ToString(), reader.GetString(0));
```
[the remark this answers is not in the document]
Comment (sfc-gh-dstempniak) [reply Snowflake.Data.Tests/IntegrationTests/SFDbCommandIT.cs:247]: uniform returns random numbers, so by definition it should not be tested, but ok, I added comment

File: Snowflake.Data/Core/BaseResultChunk.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        ARROW
+    }
+    
+    public abstract class BaseResultChunk
+    {
+        internal abstract ResultFormat Format { get; }
+        
+        public int RowCount { get; protected set; }
+        
+        public int ColCount { get; protected set; }
```
[the remark this answers is not in the document]
Comment (sfc-gh-dstempniak) [reply Snowflake.Data/Core/BaseResultChunk.cs]: I changed to ColumncCount, I agree it is better

**Comment by sfc-gh-igarish** [comment]
I saw many "public" method changes. Are we going to mark it as a BCR? Also do we added result format changes from ARROW to JSON and vice versa like "alter session set DOTNET_RESULT_FORMAT = 'ARROW' | 'JSON'"? Default should be arrow like JDBC/ODBC/GO drivers.

**Comment by sfc-gh-dstempniak** [comment]
@sfc-gh-igarish thank you for your comment. I recreated IResultChunk interface not to do BCR. Now interface of all public methods in public classes remain the same.
DOTNET_QUERY_RESULT_FORMAT parameter will be added to GS in SNOW-893840. Arrow code is already unit-tested, integration tests will be added later.

**Pull request merged**